### PR TITLE
Fix some format specs for size_t arguments

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
Salvages the fixes found with #11147

### CODE REVIEW

File: src/dmd/doc.d [hunk 1]
```diff
@@ -306,7 +306,7 @@ private final class ParamSection : Section
                             cast(int)(tf.parameterList.varargs == VarArg.variadic);
             if (pcount != paramcount)
             {
-                warning(s.loc, "Ddoc: parameter count mismatch, expected %d, got %d", pcount, paramcount);
+                warning(s.loc, "Ddoc: parameter count mismatch, expected %zd, got %zd", pcount, paramcount);
```
Comment (Geod24) [inline src/dmd/doc.d:309]: They are both `size_t` so shouldn't it be `zu` ?

File: src/dmd/dstruct.d [hunk 2]
```diff
@@ -456,7 +456,7 @@ extern (C++) class StructDeclaration : AggregateDeclaration
                     // CTFE sometimes creates null as hidden pointer; we'll allow this.
                     continue;
                 }
-                .error(loc, "more initializers than fields (%d) of `%s`", nfields, toChars());
+                .error(loc, "more initializers than fields (%zd) of `%s`", nfields, toChars());
```
Comment (Geod24) [inline src/dmd/dstruct.d:459]: nfields is also `size_t` so `zu` ? (https://github.com/dlang/dmd/blob/c6ab9c400f54704a1287facc2de03d7a10c10e33/src/dmd/aggregate.d#L251)

File: src/dmd/dsymbolsem.d [hunk 3]
```diff
@@ -2188,7 +2188,7 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
                             return; // An error happened in `identFromSE`
                     }
                     else
-                        ns.exp.error("`%s`: index %d is not a string constant, it is a `%s`",
+                        ns.exp.error("`%s`: index %zd is not a string constant, it is a `%s`",
```
Comment (Geod24) [inline src/dmd/dsymbolsem.d:2191]: `size_t`

File: src/dmd/link.d [hunk 4]
```diff
@@ -831,7 +831,7 @@ version (Windows)
                 if (status == 0)
                     args = "@_CMDLINE";
                 else
-                    error(Loc.initial, "command line length of %d is too long", len);
+                    error(Loc.initial, "command line length of %zd is too long", len);
```
Comment (Geod24) [inline src/dmd/link.d:834]: `size_t`

**Review by Geod24 — COMMENTED** [review]
Looks like in many places `zd` is used instead of `zu` for `size_t`. I'm surprised `pragma(printf)` doesn't warn about this. In any case, shouldn't they all be `size_t` ? I noted a few, but you probably need to go over all the diff.

**Comment by MoonlightSentinel** [comment]
`pragma(printf)` ignores signedness as long as the argument size match.

I just added the `z` but can also make it `zu` whenever it's appropriate.

File: src/dmd/todt.d [hunk 5]
```diff
@@ -165,7 +165,7 @@ extern (C++) void Initializer_toDt(Initializer init, ref DtBuilder dtb)
                 }
                 else if (ai.dim > tadim)
                 {
-                    error(ai.loc, "too many initializers, %d, for array[%d]", ai.dim, tadim);
+                    error(ai.loc, "too many initializers, %u, for array[%zu]", ai.dim, tadim);
```
Comment (Geod24) [inline src/dmd/todt.d:168]: `%u` for `dim ?

**Review by Geod24 — APPROVED** [review]
Just one last comment

File: src/dmd/todt.d (hunk 5 above)
Comment (MoonlightSentinel) [reply src/dmd/todt.d:168]: Nope:
https://github.com/dlang/dmd/blob/b4744b28e21cbe28a98650417492a8a79367095f/src/dmd/init.d#L164-L168

Comment (Geod24) [reply src/dmd/todt.d:168]: Sneaky

Comment (MoonlightSentinel) [reply src/dmd/todt.d:168]: Was touching that line anyway, better fix it now than later

**Review by WalterBright — CHANGES_REQUESTED** [review]
Replace use of z format with casts.

File: src/dmd/attrib.d [hunk 6]
```diff
@@ -842,7 +842,7 @@ extern (C++) final class PragmaDeclaration : AttribDeclaration
                 inlining = PINLINE.default_;
             else if (args.dim != 1)
             {
-                error("one boolean expression expected for `pragma(inline)`, not %d", args.dim);
+                error("one boolean expression expected for `pragma(inline)`, not %zu", args.dim);
```
[the remark this answers is not in the document]
Comment (MoonlightSentinel) [reply src/dmd/attrib.d]: Fixed

**Review by WalterBright — APPROVED** [review]

**Comment by Geod24** [comment]
@WalterBright : What's our support period for the host C++ compiler ? The first version that supports `z` is VS2013, meaning we are still supporting VS2010 ?

**Comment by WalterBright** [comment]
> The first version that supports z is VS2013, meaning we are still supporting VS2010 ?

I was told a month or two ago we were still supporting VS2013 in a build.

**Comment by rainers** [comment]
> The first version that supports `z` is VS2013, meaning we are still supporting VS2010 ?

No, it is only supported starting with VS2015.

> I was told a month or two ago we were still supporting VS2010 in a build.

We still support it, but that doesn't mean that dmd has to be build with it.

The release is built with VS2017 since 2.091 IIRC, Azure pipelines use that version, too. But autotester is still running VS2013.

**Pull request merged**